packages/ses/test/test-scope.js

```diff
@@ -0,0 +1,199 @@
+
+  delete globalThis.bar;
+});
+
+test('scope behavior - lookup in sloppyGlobalsMode', t => {
+  t.plan(7);
+
+  const globalObject = {};
+  const globalLexicals = {};
+  const { safeEvaluate: evaluate } = makeSafeEvaluator({
+    globalObject,
+    globalLexicals,
+    sloppyGlobalsMode: true,
+  });
+
+  globalThis.bar = {};
+
+  t.is(evaluate('globalThis'), undefined);
+  t.is(evaluate('eval'), undefined);
+  t.is(evaluate('foo'), undefined);
+  t.is(evaluate('bar'), undefined);
+  t.is(evaluate('foobar'), undefined);
+  t.is(evaluate('dummy'), undefined);
+
+  t.deepEqual(evaluate('arguments'), ['arguments']);
+
+  delete globalThis.bar;
+});
+
+test('scope behavior - this-value', t => {
+  t.plan(6);
+
+  let hogeValue;
+  let fugaValue;
+
+  const globalObject = {
+    get foo() {
+      return this;
+    },
+    set hoge(_value) {
+      hogeValue = this;
+    },
+    quux() {
+      return this;
+    },
+  };
+  const globalLexicals = {
+    get bar() {
+      return this;
+    },
+    set fuga(_value) {
+      fugaValue = this;
+    },
+    garply() {
+      return this;
+    },
+  };
+  const knownScopeProxies = new WeakSet();
+  const { safeEvaluate: evaluate } = makeSafeEvaluator({
+    globalObject,
+    globalLexicals,
+    knownScopeProxies,
+  });
+
+  t.is(evaluate('foo'), globalObject);
+  t.is(evaluate('bar'), globalObject);
+
+  evaluate('hoge = 123');
+  evaluate('fuga = 456');
+  t.is(hogeValue, globalObject);
+  t.is(fugaValue, globalObject);
+
+  t.is(knownScopeProxies.has(evaluate('quux()')), true);
+  t.is(knownScopeProxies.has(evaluate('garply()')), true);
+});
+
+test('scope behavior - assignment', t => {
+  t.plan(13);
+
+  const globalObject = { foo: {} };
+  const globalLexicals = { foobar: {} };
+  const { safeEvaluate: evaluate } = makeSafeEvaluator({
+    globalObject,
+    globalLexicals,
+  });
+  const doAssignment = (leftHandRef, value) => {
+    evaluate(`(value) => { ${leftHandRef} = value }`)(value);
+  };
+
+  globalThis.bar = {};
+
+  const evil = {};
+  // eslint-disable-next-line no-eval
+  const originalEval = globalThis.eval;
+  doAssignment('this.eval', evil);
+  t.is(globalObject.eval, evil);
+  // eslint-disable-next-line no-eval
+  t.is(globalThis.eval, originalEval);
+
+  const bar = {};
+  const originalBar = globalThis.bar;
+  doAssignment('this.bar', bar);
+  t.is(globalObject.bar, bar);
+  t.is(globalThis.bar, originalBar);
+
+  const foo = {};
+  const originalFoo = globalObject.for;
+  doAssignment('foo', foo);
+  t.is(globalObject.foo, foo);
+  t.not(globalObject.foo, originalFoo);
+  t.is(globalThis.foo, undefined);
+
+  const foobar = {};
+  const originalFoobar = globalLexicals.foobar;
+  doAssignment('foobar', foobar);
+  t.is(globalLexicals.foobar, foobar);
+  t.not(globalObject.foo, originalFoobar);
+  t.is(globalObject.foobar, undefined);
+  t.is(globalThis.foobar, undefined);
+
+  t.is(Object.keys(globalObject).length, 3);
+  t.is(Object.keys(globalLexicals).length, 1);
+
```

Comment:
`doAssignment('eval', evil)` would be an interesting case, and of course verifying that this does not break the machinery of the safe evaluator afterward.

Reply:
assigning to `eval` without first doing an assignment to `this.eval` will throw a `ReferenceError` when `sloppyGlobalsMode` is not set

Reply:
as for testing if that assignment breaks the evaluator, we are already indirectly testing that

Reply:
Works for me as long as it’s not the last assertion.